design/open/uncomitted-gc.md

```diff
@@ -0,0 +1,114 @@
+       eventually be deleted by the GC job.
+
+#### CopyObject
+
+1. Copy object in the same branch will work the same - creating a new staging entry using the existing entry information.
+2. For objects that are not part of the branch, use the underlying adapter copy operation.
+3. "Move/Rename" operation will need to use the new version of CopyObject instead of StageObject.
+
+With the above changes, bucket data is partitioned by repositories, branches and finally a bounded partition.
+This allows running the GC process at branch level, reduces the data scan size and allows parallelization.
+
+#### PrepareUncommittedForGC
+
+A new API which will create meta-ranges and ranges using the given branch uncommitted data. These files
+will be saved in a designated path used by the GC client to list branch's uncommitted objects.  
+For the purpose of this document we'll call this the `GC commit` (feel free to suggest a better term :) )
+
+#### Reading data from lakeFS
+
+Reading data from lakeFS will be similar to the current GC process - using the SSTable reader to quickly list branch objects.
+The above-mentioned API will enable doing so for all branch objects (committed and uncommitted).
+
+### GC Flows
+
+#### Flow 1: Clean Run
+
+1. Run PrepareUncommittedForGC
+2. Read all addresses from branch commits -> `lakeFS DF`
+3. Read all addresses from branch `GC commit` -> `lakeFS DF`
+4. Read all objects on branch path (can be done in parallel by 'partition') -> `Branch DF`
+5. Subtract results `lakeFS DF` from `Branch DF`
+6. Filter files newer than < TOKEN_EXPIRY_TIME > and special paths
+7. The remainder is a list of files which can be safely removed
+8. Finally, save the current run's `GC commit` the last read partition and newest commit id
+
+#### Flow 2: Optimized Run
+
+Optimized run uses the previous GC run output, to perform a partial scan of the branch to remove uncommitted garbage.
+
+##### Step 1. Analyze Data and Perform Cleanup for old entries (GC client)
+
+1. Run PrepareUncommittedForGC
+2. Read addresses from branch's new commits (all commits up to the last GC run commit id) -> `lakeFS DF`
+3. Read addresses from branch `GC commit` -> `lakeFS DF`
+4. Subtract results `lakeFS DF` from previous run's `GC commit`
+5. The result is a list of files that can be safely removed
+
+>**Note:** This step handles cases of objects that were uncommitted during previous GC run and are now deleted
+
+##### Step 2. Analyze Data and Perform Cleanup for new entries (GC client)
+1. Read all objects on branch path up to the previous run's last read partition (can be done in parallel by 'partition') -> `Branch DF`
+2. Subtract `lakeFS DF` from `Branch DF`
+3. Filter files newer than < TOKEN_EXPIRY_TIME > and special paths
+4. The remainder is a list of files which can be safely removed
+5. Finally, save the current run's `GC commit` the last read partition and newest commit id
+
+## Limitations
```

Comment:
It's also a dedicated flow per branch:
1. How do I discover branches that were deleted?
2. How do I skip old branches that were cleaned up and not used since?
3. Branch discovery (I assume `listBranches` should suffice..?)

Reply:
Answer 3: Using list branches should suffice
Answer 1: Listing the `branches` prefix and comparing with the above should solve the issue

Answer 2: This might be a topic for "phase 2" or optimizations effort. But consider the following assumptions:
1. No new partitions were created on branch
2. No new commits on the branch
3. Uncommitted list equals previous run's uncommitted
It is simply the process of listing the last branch partition and comparing it with the branch's current uncommitted objects which should be a _relatively_ lightweight operation